conda_build/index.py

```diff
@@ -5,78 +5,155 @@
+        mtime = os.path.getmtime(index_file)
 
+    if (clear_cache or
+            not os.path.isfile(index_file) or
+            local_subdir != subdir or
+            mtime > local_index_timestamp or
+            cached_channels != channel_urls):
+
+        log.debug("Building new index for subdir '{}' with channels {}, condarc channels "
+                  "= {}".format(subdir, channel_urls, not omit_defaults))
+        # priority: local by croot (can vary), then channels passed as args,
+        #     then channels from config.
+        capture = contextlib.contextmanager(lambda: (yield))
+        if debug:
+            log_context = partial(utils.LoggingContext, logging.DEBUG)
+        elif verbose:
+            log_context = partial(utils.LoggingContext, logging.WARN)
+        else:
+            log_context = partial(utils.LoggingContext, logging.CRITICAL + 1)
+            capture = utils.capture
+
+        urls = list(channel_urls)
+        if os.path.isdir(output_folder):
+            urls.insert(0, url_path(output_folder))
+        _ensure_valid_channel(output_folder, subdir, verbose=verbose, locking=locking,
+                              timeout=timeout)
+
+        # silence output from conda about fetching index files
+        with log_context():
+            with capture():
+                # replace noarch with native subdir - this ends up building an index with both the
+                #      native content and the noarch content.
+                if subdir == 'noarch':
+                    subdir = conda_interface.subdir
+                try:
+                    cached_index = get_index(channel_urls=urls,
+                                    prepend=not omit_defaults,
+                                    use_local=False,
+                                    use_cache=False,
+                                    platform=subdir)
+                # HACK: defaults does not have the many subfolders we support.  Omit it and
+                #          try again.
+                except CondaHTTPError:
+                    if 'defaults' in urls:
+                        urls.remove('defaults')
+                    cached_index = get_index(channel_urls=urls,
+                                             prepend=omit_defaults,
+                                             use_local=False,
+                                             use_cache=False,
+                                             platform=subdir)
+        local_index_timestamp = os.path.getmtime(index_file)
+        local_subdir = subdir
+        cached_channels = channel_urls
+    return cached_index, local_index_timestamp
 
-def _add_extra_path(extra_paths, path):
-    if isfile(path):
-        extra_paths[basename(path)] = {
-            'size': getsize(path),
-            'timestamp': int(getmtime(path)),
-            'md5': md5_file(path),
-        }
+
+def _ensure_valid_channel(local_folder, subdir, verbose=True, locking=True, timeout=90):
```

Comment:
This function is not modified in this PR, just shifted to a new location in the file.  No need to review.